src/main/java/io/netty/handler/codec/compression/BrotliDecoder.java

```diff
@@ -0,0 +1,111 @@
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+/**
+ * Decompress a {@link ByteBuf} using the inflate algorithm.
+ */
+public class BrotliDecoder extends ByteToMessageDecoder {
+
+  private static final InternalLogger log =
+      InternalLoggerFactory.getInstance(BrotliDecoder.class);
+
+  static {
+    BrotliLoader.isBrotliAvailable();
+  }
+
+  /*
+  For how this value is derived, please see: `BROTLI_MAX_NUMBER_OF_BLOCK_TYPES` in these docs:
+     - https://github.com/google/brotli/blob/master/c/common/constants.h
+     - https://tools.ietf.org/html/draft-vandevenne-shared-brotli-format-01
+   */
+  private static int BROTLI_MAX_NUMBER_OF_BLOCK_TYPES = 256;
+
+  public BrotliDecoder() {
+  }
+
+  public static byte[] decompress(byte[] compressedArray) throws IOException {
+    if(compressedArray == null) {
+      return null;
+    }
+
+    ByteArrayOutputStream out;
+    try (BrotliInputStream is = new BrotliInputStream(new ByteArrayInputStream(compressedArray))) {
+      out = new ByteArrayOutputStream();
+      if (!decompress(out, is)) {
+        return null;
+      }
+    } catch (IOException e) {
+      log.error("Unhandled exception when decompressing brotli", e);
+      throw e;
+    }
+    return out.toByteArray();
+  }
+
+  @Override
+  protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+    /*
+       use in.alloc().buffer() instead of Unpooled.buffer() as best practice.
+       See: https://github.com/netty/netty/wiki/New-and-noteworthy-in-4.0#pooled-buffers
+    */
+    try (ByteBufOutputStream output = new ByteBufOutputStream(in.alloc().buffer())) {
+      try (BrotliInputStream brotliInputStream = new BrotliInputStream(new ByteBufInputStream(in))) {
+        if (!decompress(output, brotliInputStream)) return;
+      } catch (IOException e) {
+        log.error("Unhandled exception when decompressing brotli", e);
+        throw e;
+      }
+      out.add(output.buffer());
+    }
+  }
+
+  private static boolean decompress(OutputStream output, BrotliInputStream brotliInputStream) throws IOException {
+    byte[] decompressBuffer = new byte[BROTLI_MAX_NUMBER_OF_BLOCK_TYPES];
+    // is the stream ready for us to decompress?
+    int bytesRead;
+    try {
+      bytesRead = brotliInputStream.read(decompressBuffer);
+    } catch (IOException e) {
+      // unexpected end of input, not ready to decompress, so just return
+      return false;
+    }
+    // continue reading until we have hit EOF
+    while (bytesRead > -1) { // -1 means EOF
+      output.write(decompressBuffer, 0, bytesRead);
+      Arrays.fill(decompressBuffer, (byte) 0);
+      bytesRead = brotliInputStream.read(decompressBuffer);
+    }
+    return true;
+  }
```

Comment:
```suggestion
  private static boolean decompress(OutputStream output, BrotliInputStream brotliInputStream) {
    byte[] decompressBuffer = new byte[BROTLI_MAX_NUMBER_OF_BLOCK_TYPES];
    // is the stream ready for us to decompress?
    int bytesRead;
    try {
      bytesRead = brotliInputStream.read(decompressBuffer);
      // continue reading until we have hit EOF
      while (bytesRead > -1) { // -1 means EOF
        output.write(decompressBuffer, 0, bytesRead);
        Arrays.fill(decompressBuffer, (byte) 0);
        bytesRead = brotliInputStream.read(decompressBuffer);
      }
      return true;
    } catch (IOException e) {
      log.warn("Could not decompress a brotli stream.", e);
      // unexpected end of input, not ready to decompress, so just return
      return false;
    } catch (RuntimeException e) {
      log.warn("Could not decompress a brotli stream.", e);
      // unexpected runtime error
      return false;
    }
  }
```

Reply:
this impl has an uneven error handling strategy.

Reply:
changes the impl so
1 - the error handling strategy is consistent
2 - runtime errors coming from the netty byte buf manipulations are accounted for
3 - removed the declared checked IOException from the signature.

Reply:
> the netty byte buf manipulations

https://github.com/nixxcode/jvm-brotli/blob/f4e21fd3e5a321a65e670d28beb3931bb6ffbda6/jvmbrotli/src/main/java/com/nixxcode/jvmbrotli/dec/BrotliInputStream.java#L77-L96